pallets/vec-set/src/lib.rs

```diff
@@ -25,71 +31,84 @@ decl_event!(
-		/// This method is faster, and therefore preferred, whenever possible
+		/// Adds a member to the membership set unless the max is reached
 		#[weight = 10_000]
-		fn append_new_entries(origin) -> DispatchResult {
-			let user = ensure_signed(origin)?;
-
-			// this encodes the new values and appends them to the already encoded existing evc
-			Self::new_values()
-				.iter()
-				.for_each(CurrentValues::append);
-			Self::deposit_event(RawEvent::AppendVec(user));
-			Ok(())
-		}
-
-		#[weight = 10_000]
-		fn add_member(origin) -> DispatchResult {
+		pub fn add_member(origin) -> DispatchResult {
 			let new_member = ensure_signed(origin)?;
-			ensure!(!Self::is_member(&new_member), "must not be a member to be added");
-			<Members<T>>::append(new_member.clone());
-			Self::deposit_event(RawEvent::MemberAdded(new_member));
-			Ok(())
+
+			let mut members = Members::<T>::get();
+			ensure!(members.len() < MAX_MEMBERS, Error::<T>::MembershipLimitReached);
+
+			// We don't want to add duplicate members, so we check whether the potential new
+			// member is already present in the list. Because the list is always ordered, we can
+			// leverage the binary search which makes this check O(log n).
+			match members.binary_search(&new_member) {
+				// If the search succeeds, the caller is already a member, so just return
+				Ok(_) => Err(Error::<T>::AlreadyMember.into()),
+				// If the search fails, the caller is not a member and we learned the index where
+				// they should be inserted
+				Err(index) => {
+					members.insert(index, new_member.clone());
+					Members::<T>::put(members);
+					Self::deposit_event(RawEvent::MemberAdded(new_member));
+					Ok(())
+				}
+			}
 		}
 
+		/// Removes a member.
 		#[weight = 10_000]
 		fn remove_member(origin) -> DispatchResult {
 			let old_member = ensure_signed(origin)?;
-			ensure!(Self::is_member(&old_member), "must be a member in order to leave");
-			<Members<T>>::mutate(|v| v.retain(|i| i != &old_member));
-			Self::deposit_event(RawEvent::MemberRemoved(old_member));
-			Ok(())
+
+			let mut members = Members::<T>::get();
+
+			// We have to find out if the member exists in the sorted vec, and, if so, where.
+			match members.binary_search(&old_member) {
+				// If the search succeeds, the caller is a member, so remove her
+				Ok(index) => {
+					members.remove(index);
+					Members::<T>::put(members);
+					Self::deposit_event(RawEvent::MemberRemoved(old_member));
+					Ok(())
+				},
+				// If the search fails, the caller is not a member, so just return
+				Err(_) => Err(Error::<T>::NotMember.into()),
+			}
 		}
+
 		// also see `append_or_insert`, `append_or_put` in pallet-elections/phragmen, democracy
 	}
 }
 
-impl<T: Trait> Module<T> {
-	pub fn is_member(who: &T::AccountId) -> bool {
-		<Members<T>>::get().contains(who)
+impl<T: Trait> AccountSet for Module<T> {
+	type AccountId = T::AccountId;
+
+	fn accounts() -> BTreeSet<T::AccountId> {
```

Comment:
It would totally work, and I considered it for quite a while.

While it makes the implementation for `vec-set` much nicer, it makes the implementation for `map-set` worse. In particular, when iterating the map, we would have to either manually perform the binary search to keep the vec sorted or give up the sorting, entirely.

I'm interested to learn more about the cloning though. I guess the cloning happens because I used `into_iter`? Is that avoidable?

Reply:
> I'm interested to learn more about the cloning though. I guess the cloning happens because I used `into_iter`? Is that avoidable?

Hmm, maybe I'm wrong. `Self::members()` returns an owned value and this code consumes that `Vec` so the `IntoIter` impl likely just takes owned values and hands them to `collect()` which use the [`FromIter`](https://doc.rust-lang.org/src/alloc/collections/btree/map.rs.html#1763-1769) impl for `BTreeSet`.
I thought the `BTreeSet` impl couldn't re-use those values when creating the new `BTreeSet` but [I could be wrong about that](https://doc.rust-lang.org/src/alloc/collections/btree/map.rs.html#2296-2333). I actually don't know!

Rust is smarter than me, who knew! :D